### PR TITLE
Support for enabling or disabling a test on a range of JREs

### LINKED ISSUE #2016: Allow tests to be enabled or disabled on a range of JREs
I'd like to test some code that behaves in different ways depending on the version of the JRE on which it's running. In my particular case, the code behaves in one way on Java 12 and earlier and in another on Java 13 and later. We have CI jobs that test on Java 8, 11, 12, and 13 (we'll probably drop 12 soon due to it being superseded by 13).

As CI jobs on new JREs are added and removed, I'd prefer for it not to be necessary to change the code to make sure that the right tests are enabled. As far as I can tell, even if `@EnabledOnJre` and `@DisabledOnJre` were repeatable, this goal would not be met. Instead, what I'd really like to be able to do is specify a (possibly open-ended) range of JREs on which a test should be enabled or disabled. In my particular case being able to enable a test on JRE 12 and earlier and another on JRE 13 and later would be sufficient, but I'm sure there are other use cases that would benefit from support for closed ranges.

## Deliverables

- [ ] Support for enabling or disabling a test on a range of JREs

### PR DESCRIPTION
**Pull request by aepfli**
## Overview

Support for enabling or disabling a test on a range of JREs. Therfore we implemented two new annotations `@EnabledForJreRange` and `@DisabledForJreRange` which contain two parameters `min` (Defaults to `JRE.Java_8`)  and `max` (defaults to `JRE.OTHER`). By defaulting those to the highest and lowest version, we allow the usage of half open ranges.

I might need to rework javadoc and documentation, i am not really good when it comes to wording those.

Resolves #2016.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by aepfli** [comment]
setting up JDK 14 locally, to check the problem

**Comment by sormuras** [comment]
```
org.junit.jupiter.api.condition.EnabledForJreRangeConditionTests > javaMin10() FAILED
    org.opentest4j.AssertionFailedError: Should be disabled ==> expected: <true> but was: <false>
        at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
        at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:40)
        at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:193)
        at org.junit.jupiter.api.condition.AbstractExecutionConditionTests.assertDisabled(AbstractExecutionConditionTests.java:58)
        at org.junit.jupiter.api.condition.EnabledForJreRangeConditionTests.assertEnabledOnCurrentJreIf(EnabledForJreRangeConditionTests.java:124)
        at org.junit.jupiter.api.condition.EnabledForJreRangeConditionTests.javaMin10(EnabledForJreRangeConditionTests.java:105)

org.junit.jupiter.api.condition.DisabledForJreRangeConditionTests > javaMin10() FAILED
    org.opentest4j.AssertionFailedError: Should be enabled ==> expected: <true> but was: <false>
        at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
        at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:40)
        at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:193)
        at org.junit.jupiter.api.condition.AbstractExecutionConditionTests.assertEnabled(AbstractExecutionConditionTests.java:54)
        at org.junit.jupiter.api.condition.DisabledForJreRangeConditionTests.assertDisabledOnCurrentJreIf(DisabledForJreRangeConditionTests.java:124)
        at org.junit.jupiter.api.condition.DisabledForJreRangeConditionTests.javaMin10(DisabledForJreRangeConditionTests.java:105)
```

Interesting!

**Comment by aepfli** [comment]
yeah stupid me, did not add java14 to the test when we just define a lower boundary. i just missed a `|| onJava14()` in the condition. The thing which is troubling me, would be that this test case always needs manual adaption, as soon as another JRE is added. which is in my opinion not a good practice.

Test is running locally now, should i add there some kind of comment or remove it? Everytime there is a new JDK the upper bound will change, and will force developers to adapt those 2 tests.

**Comment by sormuras** [comment]
That's why I (alone) favor an `int`-based annotation value approach. Here, we could use `Integer.MAX_VALUE` for the upper limit AND be independent from the `JRE` enum.

I guess, that location (the test with the added `|| onJava14()`) is another place that has to be touched as soon as the `JRE` enum is extended. Twice a year.

**Comment by aepfli** [comment]
or i do
```
assertEnabledOnCurrentJreIf(!(onJava8() || onJava9()))

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledForJreRangeCondition.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+				.map(disabledForJreRange -> {
+					JRE min = disabledForJreRange.min();
+					JRE max = disabledForJreRange.max();
+					Preconditions.condition(max.compareTo(min) >= 0,
+						"@DisabledForJreRange.min must be lower or the same version as @DisabledForJreRange.max");
+
+					if (isCurrentVersionWithinRange(min, max)) {
+						return DISABLED_ON_CURRENT_JRE;
+					}
+					else {
```
[the remark this answers is not in the document]
Comment (aepfli) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledForJreRangeCondition.java]: just to be sure i understand the terminology correct:
```
if (isCurrentVersionWithinRange(min, max)) {
	return ENABLED_ON_CURRENT_JRE;
}
return DISABLED_ON_CURRENT_JRE;
```
this is what you meant by frame :D

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/JRE.java [hunk 2]
```diff
@@ -96,7 +98,7 @@
 
 	private static final Logger logger = LoggerFactory.getLogger(JRE.class);
 
-	private static final JRE CURRENT_VERSION = determineCurrentVersion();
+	public static final JRE CURRENT_VERSION = determineCurrentVersion();
```
[the remark this answers is not in the document]
Comment (aepfli) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/JRE.java]: what would be your preference?

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledForJreRangeCondition.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sormuras) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledForJreRangeCondition.java]: Exactly.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/JRE.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sormuras) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/JRE.java]: Package private sounds like the best option to me.

**Comment by aepfli** [comment]
> Also, please also add a test case the combines a JRE range with explicit `@DisabledOnJre()`, like:
> 
> ```java
> @EnabledForJreRange(min = JAVA_9, max = JAVA_13)
> @DisableOnJre({JAVA_10, JAVA_12})
> void testEnabledOn_9_11_13() {}
> ```

I am not sure, if i should tackle this within this pull request, because i think i will need to change ` `org.junit.jupiter.api.condition.AbstractExecutionConditionTests`. As right now it only supports one Condition to be evaluated. Yes i do see the benefit of this, but before investigating work into that, which might not fit the purpose, i thought it is better to discuss it. So what options do i have?
- changing `AbstractExectuionTests` to work with multiple `Conditions`
- as this is the only case, making the `AbstractExecutionContext.context` package private and overwriting the methods for evaluation and verification. i still would need to implement the abstract method `getExecutionCondition` but just with a fake implementation
- leave it as is, as so far there have not been any "combined" JRE Annotation tests
- something i did not think of :D

Thank you

**Comment by sormuras** [comment]
It should work right out of the box - as `@EnabledForJreRange` and `@DisableOnJre` are distinct conditional annotations, this [note](https://junit.org/junit5/docs/current/user-guide/#writing-tests-conditional-execution) comes into play: _Note, however, that each conditional annotation may be used in conjunction with other conditional annotations in the org.junit.jupiter.api.condition package._

**Comment by aepfli** [comment]
> It should work right out of the box - as `@EnabledForJreRange` and `@DisableOnJre` are distinct conditional annotations, this [note](https://junit.org/junit5/docs/current/user-guide/#writing-tests-conditional-execution) comes into play: _Note, however, that each conditional annotation may be used in conjunction with other conditional annotations in the org.junit.jupiter.api.condition package._

yes i am aware of that, but the current tests for conditions is not build to evaluate two at once - Not the general evaluation, but the *TEST* evaluation is build for one condition to evaluate

https://github.com/junit-team/junit5/blob/d6c023e10349109d7e57d642b4191e2ba5e156fd/junit-jupiter-engine/src/test/java/org/junit/jupiter/api/condition/AbstractExecutionConditionTests.java#L45-L51

**Comment by sormuras** [comment]
Ah, I see. Yes, let's keep that _TEST_ feature for later. Or as you phrased it:

> leave it as is, as so far there have not been any "combined" JRE Annotation tests

**Comment by sormuras** [comment]
Perhaps, add the combined usage as a sample to `ConditionalTestExecutionDemo`? 🤔

**Comment by aepfli** [comment]
> Perhaps, add the combined usage as a sample to `ConditionalTestExecutionDemo`?

Maybe we should also rethink of wording of the docs, because actually if a condition is enabled is nice to know, but has no impact, as current implementation is only checking if there is any condition which is disabled in the context. Maybe we want to highlight that somewhere, because i can think that people easily would assume that

```
@DisabledForJreRange(min = JAVA_9, max = JAVA_13)
@EnabledOnJre({JAVA_10, JAVA_12})
void testEnabledOn_10_112() {}
```
would work, but infact it will not. As far as i could see, nothing is mentioned in the condition java doc, nor in the ascii doc. nor maybe i am just blind :see_no_evil: 

should i additionally add my case to the ConditionalTestExecutionDemo with a comment that this is no working? or should i add my remarks to the docs for conditional tests?

**Comment by sormuras** [comment]
> Maybe we should also rethink of wording of the docs, [...]

Yeah, true. But that's not a specific of the JRE-based conditions, that applies to all **"2.7. Conditional Test Execution"** annotations and extensions. So, let's tackle that in a new issue/PR. 😉 

Note, in **"5.3. Conditional Test Execution"** it reads: _"When multiple `ExecutionCondition` extensions are registered, a container or test is disabled as soon as one of the conditions returns disabled."_ Perhaps we should rephrase this and also include it in **2.7**.

> Should i additionally add my case to the `ConditionalTestExecutionDemo` with a comment that this is not working? 

That's up to you. Better to have this one example with a comment than having it not at all.

Also, please rebase this branch and resolve the conflicts. Merci!

**Comment by sormuras** [comment]
Resolved merge conflict using the web editor ... let's see what the CI run says.

**Comment by sormuras** [comment]
Thanks @aepfli!

PRs with improvements of the general documentation and more demos/examples with explanations are always welcome.

**Comment by aepfli** [comment]
Wow that was fast now, thank you for resolving merge conflicts, etc.

Improving docs sounds like an easy win for hacktober, if I find the time, for sure.

**Pull request merged**